crates/rpc/rpc/src/debug.rs

```diff
@@ -360,6 +360,92 @@ where
         Ok(frame.into())
     }
 
+    //TODO: Add docs
+    pub async fn debug_trace_call_many(
+        &self,
+        bundles: Vec<(CallRequest, Option<GethDebugTracingCallOptions>)>,
+        block_id: Option<BlockId>,
+        transaction_index: Option<isize>,
+    ) -> EthResult<Vec<GethTrace>> {
+        if bundles.is_empty() {
+            return Err(EthApiError::InvalidParams(String::from("bundles are empty.")));
+        }
+
+        let transaction_index = transaction_index.unwrap_or(-1);
+        let state_at = block_id.unwrap_or(BlockId::Number(BlockNumberOrTag::Latest));
+        //TODO: Figure out what to do with -1 properly
+
+        let block_hash = self
+            .inner
+            .provider
+            .block_hash_for_id(state_at)?
+            .ok_or_else(|| EthApiError::UnknownBlockNumber)?;
+
+        let ((cfg, block_env, _), block) = futures::try_join!(
+            self.inner.eth_api.evm_env_at(block_hash.into()),
+            self.inner.eth_api.block_by_id(state_at),
+        )?;
+
+        let block = block.ok_or_else(|| EthApiError::UnknownBlockNumber)?;
+
+        self.on_blocking_task(|this| async move {
+            this.inner.eth_api.with_state_at_block(state_at, |state| {
+                let gas_limit = this.inner.eth_api.call_gas_limit();
+                let mut results = Vec::with_capacity(bundles.len());
+                let mut db = SubState::new(State::new(state));
+
+                let mut transactions = if transaction_index == -1 {
+                    block.body.into_iter()
+                } else {
+                    block.body
+                        [0..usize::try_from(transaction_index).expect("Tx index outside bounds.")]
+                        .to_vec()
+                        .into_iter()
+                };
```

Comment:
this could be simplified by using `.take()` when iterating over the transactions instead, for example `-1` would be `block.body.len()` == all transactions